### PR TITLE
Add support for named tuples to pretty

### LINKED ISSUE #2012: [REQUEST] Add explicit support for NamedTuples
The Pretty class doesn't support pretty printing NamedTuples.

We should attempt to auto detect named tuples in order to expand these on to multiple lines as appropriate.

NamedTuples inherit from `tuple` and don't have a common base class. They have a `_fields` attribute which could identify them as a named tuple. 

We will also need to check if a custom __repr__ has been provided (in the same was as we do dataclasses).

### PR DESCRIPTION
**Pull request by darrenburns**
## Type of changes

- [ ] Bug fix
- [x] New feature
- [ ] Documentation / docstrings
- [ ] Tests
- [ ] Other

## Checklist

- [x] I've run the latest [black](https://github.com/psf/black) with default args on new code.
- [x] I've updated CHANGELOG.md and CONTRIBUTORS.md where appropriate.
- [x] I've added tests for new code.
- [x] I accept that @willmcgugan may be pedantic in the code review.

## Description

Closes #2012 

Named tuples will be formatted similarly to dataclasses when formatted using the `pretty` module. The code for handling this is very similar to dataclasses, but I opted not to try and factor it out as I think it will reduce readability.

### CODE REVIEW

File: rich/pretty.py [hunk 1]
```diff
@@ -731,7 +753,23 @@ def iter_attrs() -> Iterable[
                     append(child_node)
 
                 pop_visited(obj_id)
-
+        elif _is_namedtuple(obj):
```
Comment (willmcgugan) [inline rich/pretty.py:756]: We are also going to need a check to see if the default `__repr__` was overriden. If it was, Rich should respect that. Have a look at `_is_dataclass_repr`, maybe we can do something similar for NamedTuple.

File: rich/pretty.py [hunk 2]
```diff
@@ -524,6 +524,28 @@ def __str__(self) -> str:
             )
 
 
+def _is_namedtuple(obj: Any) -> bool:
```
Comment (willmcgugan) [inline rich/pretty.py:527]: I wonder if we can be more precise than this. I know there is no `is_namedtuple` function, alas.

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Mainly just detecting if `__repr__` has been overriden.

File: rich/pretty.py [hunk 3]
```diff
@@ -79,6 +79,24 @@ def _is_dataclass_repr(obj: object) -> bool:
...
+
+    Args:
+        obj (object): A namedtuple
+
+    Returns:
+        bool: True if the default repr is used, False if there's a custom repr.
+    """
+    obj_file = inspect.getfile(obj.__repr__)
+    default_repr_file = inspect.getfile(_dummy_namedtuple.__repr__)
+    print(obj_file, default_repr_file)
```
Comment (willmcgugan) [inline rich/pretty.py:96]: Debug print snuck in there.

File: rich/pretty.py [hunk 4]
```diff
@@ -524,6 +543,28 @@ def __str__(self) -> str:
...
+    there is only a minuscule chance of this happening unintentionally.
+
+    Args:
+        obj (Any): The object to test
+
+    Returns:
+        bool: True if the object is a namedtuple. False otherwise.
+    """
+    base_classes = getattr(type(obj), "__bases__", [])
+    if len(base_classes) != 1 or base_classes[0] != tuple:
```
Comment (willmcgugan) [inline rich/pretty.py:558]: You can extend tuples, and named tuples. Perhaps this should just be `if not isinstance(tuple)` ?

File: rich/pretty.py [hunk 5]
```diff
@@ -524,6 +543,28 @@ def __str__(self) -> str:
...
+        return False
+
+    fields = getattr(obj, "_fields", None)
+    if not fields or not isinstance(fields, tuple):
```
Comment (willmcgugan) [inline rich/pretty.py:562]: `_fields` *could* be an empty tuple here. For the strange edge-case of a named tuple with no fieds.

Perhaps just return `isinstance(fields, tuple)` ?

File: rich/pretty.py [hunk 6]
```diff
@@ -731,7 +772,24 @@ def iter_attrs() -> Iterable[
...
+            if reached_max_depth:
+                node = Node(value_repr="...")
+            else:
+                node = Node(
+                    open_brace=f"{obj.__class__.__name__}(",
+                    close_brace=")",
+                    children=children,
+                )
+                for last, (key, value) in loop_last(obj._asdict().items()):
+                    print(last, key, value)
```
Comment (willmcgugan) [inline rich/pretty.py:787]: print statement...

File: rich/pretty.py [hunk 7]
```diff
@@ -79,6 +79,24 @@ def _is_dataclass_repr(obj: object) -> bool:
...
+def _has_default_namedtuple_repr(obj: object) -> bool:
+    """Check if an instance of namedtuple contains the default repr
+
+    Args:
+        obj (object): A namedtuple
+
+    Returns:
+        bool: True if the default repr is used, False if there's a custom repr.
+    """
+    obj_file = inspect.getfile(obj.__repr__)
```
Comment (willmcgugan) [inline rich/pretty.py:94]: This can throw a `TypeError` according to the docs. Is that an issue here?

**Review by willmcgugan — CHANGES_REQUESTED** [review]
Introspection is tricky

File: rich/pretty.py (hunk 7 above)
Comment (darrenburns) [reply rich/pretty.py]: I've handled `TypeError` and `OSError` now.

**Review by willmcgugan — APPROVED** [review]
LGTM

**Pull request merged**